src/v/v8_engine/script_dispatcher.h

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "model/metadata.h"
+
+#include <seastar/core/condition-variable.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/shared_ptr.hh>
+
+#include <absl/container/node_hash_map.h>
+#include <absl/container/node_hash_set.h>
+#include <v8_engine/script.h>
+
+namespace v8_engine {
+
+struct script_compile_params {
+    script_compile_params(ss::sstring function_name, iobuf code)
+      : function_name(std::move(function_name))
+      , code(std::move(code)) {}
+
+    ss::sstring function_name;
+    iobuf code;
+    size_t max_heap_size_bytes{10000};
+    size_t run_timeout_ms{10000};
+};
+
+template<typename Executor>
+class script_dispatcher {
+public:
+    explicit script_dispatcher(Executor& executor)
+      : _executor(executor) {}
+
+    // Insert only put compilation params to map. It must not contain value for
+    // topic.
+    bool insert(
+      const model::topic_namespace& topic,
+      ss::sstring function_name,
+      iobuf code) {
+        return _scripts
+          .emplace(
+            topic,
+            script_compile_params(std::move(function_name), std::move(code)))
+          .second;
+    }
+
+    // Remove only deletes script if it has been compiled or it notify waiters
+    // that script is removed
+    bool remove(const model::topic_namespace& topic) {
+        auto it = _scripts.find(topic);
+        if (it == _scripts.end()) {
+            return false;
+        }
+
+        ss::visit(
+          it->second,
+          [](const script_compile_params&) {
+              // Do nothing, because nobody try to get this script
+              return;
+          },
+          [](ss::lw_shared_ptr<script_wrpapper> script_ptr) {
+              if (script_ptr->is_compiling()) {
+                  script_ptr->destroy();
+              }
+              return;
+          });
+
+        _scripts.erase(it);
+
+        return true;
+    }
+
+    ss::future<ss::lw_shared_ptr<script>> get(model::topic_namespace topic) {
+        auto it = _scripts.find(topic);
+        if (it == _scripts.end()) {
+            co_return nullptr;
+        }
+
+        co_return co_await ss::visit(
```

Comment:
is the primary reason for the complexity here that you are supporting the `insert` method which requires that you keep some code around before the runtime that uses that code is built? if so, perhaps we can simplify things.

What would prevent the script dispatcher from either (1) being provided the code by the callers of get or (2) being able to interact with other sub-systems to request the code on-demand? I think that (2) makes the most sense.

Is there a reason why we need to support `insert` and the arrival of code before that code is requested via `get`?

Reply:
In my head this order (deploy script, insert data_policy) is only one way. But I agree it is strange, and rewrote it.